--- python-openzwave-20120828.orig/openzwave.pyx	2012-08-30 00:42:44.000000000 +0200
+++ python-openzwave-20120828//openzwave.pyx	2012-08-30 00:42:51.000000000 +0200
@@ -329,7 +329,7 @@
     EnumWithDoc('Button').setDoc(   "A write-only value that is the equivalent of pressing a button to send a command to a device"),
     ]
 
-cdef map[uint32, ValueID] values_map 
+cdef map[uint64, ValueID] values_map 
 
 cdef addValueId(ValueID v, n):
     cdef string value
@@ -353,7 +353,7 @@
                     'readOnly': manager.IsValueReadOnly(v),
                     }   
     
-    values_map.insert ( pair[uint32, ValueID] (v.GetId(), v)) 
+    values_map.insert ( pair[uint64, ValueID] (v.GetId(), v)) 
 
 cdef void callback(const_notification _notification, void* _context) with gil:
     cdef Notification* notification = <Notification*>_notification
