--- python-openzwave-20120828.orig/openzwave.pyx	2012-08-30 22:03:55.000000000 +0200
+++ python-openzwave-20120828//openzwave.pyx	2012-08-30 22:14:22.000000000 +0200
@@ -17,7 +17,7 @@
 ctypedef unsigned int uint32
 ctypedef unsigned long uint64
 ctypedef int int32
-ctypedef int int16
+ctypedef short int16
 ctypedef unsigned char uint8
 
 cdef extern from "Options.h" namespace "OpenZWave":
@@ -1229,13 +1229,12 @@
             elif datatype == "Decimal":
                 type_float = value
                 cret = self.manager.SetValue(values_map.at(id), type_float)
-            # TODO: this gives me an "ambiguous overloaded method", I don't understand why.
-            #elif datatype == "Int":
-            #    type_int = value
-            #    cret = self.manager.SetValue(values_map.at(id), int32(type_int))
-            #elif datatype == "Short":
-            #    type_short = value
-            #    cret = self.manager.SetValue(values_map.at(id), type_short)
+            elif datatype == "Int":
+                type_int = value
+                cret = self.manager.SetValue(values_map.at(id), type_int)
+            elif datatype == "Short":
+                type_short = value
+                cret = self.manager.SetValue(values_map.at(id), type_short)
             elif datatype == "String":
                 type_string = string(value)
                 cret = self.manager.SetValue(values_map.at(id), type_string)
