diff -Nru python-openzwave-20120508.orig//setup.py python-openzwave-20120508//setup.py
--- python-openzwave-20120508.orig//setup.py	2012-08-10 23:45:11.000000000 +0200
+++ python-openzwave-20120508//setup.py	2012-08-10 23:46:23.000000000 +0200
@@ -9,13 +9,13 @@
                              language="c++",
                              extra_objects=['openzwave/cpp/lib/windows-mingw32/openzwave.a'],
                              include_dirs=['openzwave/cpp/src', 'openzwave/cpp/src/value_classes', 'openzwave/cpp/src/platform']
-    )]    
+    )]
 else:
     ext_modules = [extension.Extension("openzwave", ["openzwave.pyx"],
-                             libraries=['udev', 'stdc++'],
+                             libraries=['udev', 'stdc++', 'openzwave'],
                              language="c++",
-                             extra_objects=['openzwave/cpp/lib/linux/openzwave.a'], 
-                             include_dirs=['openzwave/cpp/src', 'openzwave/cpp/src/value_classes', 'openzwave/cpp/src/platform']
+                             extra_objects=['/usr/lib/libopenzwave.a'],
+                             include_dirs=['/usr/include/openzwave', '/usr/include/openzwave/value_classes', '/usr/include/openzwave/platform']
     )]
 
 setup(
@@ -25,4 +25,4 @@
   url='http://projects.maartendamen.com/projects/pyopenzwave/',
   cmdclass = {'build_ext': build_ext},
   ext_modules = ext_modules
-)
\ Pas de fin de ligne à la fin du fichier.
+)
